units/include/scipp/units/dimension.h

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2019 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#ifndef SCIPP_UNITS_DIMENSION_H
+#define SCIPP_UNITS_DIMENSION_H
+
+#include <cstdint>
+
+namespace scipp::units {
+
+/// Macro to declare available dimension labels in a particular namespace.
+///
+/// The arguments passed to the macro are converted into enum labels. `Invalid`
+/// is a reserved name and is automatically added to the list. This also defines
+/// a `to_string` function for the defined enum.
+#define SCIPP_UNITS_DECLARE_DIMENSIONS(...)                                    \
+  enum class Dim : uint16_t { __VA_ARGS__, Invalid };                          \
+                                                                               \
+  namespace detail2 {                                                          \
+  constexpr const char *names = #__VA_ARGS__;                                  \
```

Comment:
Does `constexpr const` do anything more than just `constexpr`?

Reply:
Yes, I was confused by that too, see https://stackoverflow.com/questions/30561104/const-constexpr-char-vs-constexpr-char. "constexpr applies to the pointer, making it const, but not the object that it points to. So this is one of a few cases where you might need to combine const and constexpr"